### >>> Global Setting
# - human H3K9me3 peaks
hs_c4_93_pks=/home/cmq/bioinfo/project-cmq/embryo_93/results/idr/p_0.05/filtered/h4C93-Ip-1_combined_R1_val_1_h4C93-Ip-1_combined_R2_val_2_psorted_dedup_peaks_h4C93-Ip-2_combined_R1_val_1_h4C93-Ip-2_combined_R2_val_2_psorted_dedup_peaks_ucsc_bl.bed
hs_c8_93_pks=/home/cmq/bioinfo/project-cmq/embryo_93/results/idr/p_0.05/filtered/h8C93-Ip-3_combined_R1_val_1_h8C93-Ip-3_combined_R2_val_2_psorted_dedup_peaks_h8C93-Ip-4_combined_R1_val_1_h8C93-Ip-4_combined_R2_val_2_psorted_dedup_peaks_ucsc_bl.bed
hs_morula_93_pks=/home/cmq/bioinfo/project-cmq/embryo_93/results/idr/p_0.05/filtered/hM93-Ip-1_combined_R1_val_1_hM93-Ip-1_combined_R2_val_2_psorted_dedup_peaks_hM93-Ip-2_combined_R1_val_1_hM93-Ip-2_combined_R2_val_2_psorted_dedup_peaks_ucsc_bl.bed
hs_icm_93_pks=/home/cmq/bioinfo/project-cmq/embryo_93/results/idr/p_0.05/filtered/hICM93-IP-1_combined_R1_val_1_hICM93-IP-1_combined_R2_val_2_psorted_dedup_peaks_hICM93-IP-2_combined_R1_val_1_hICM93-IP-2_combined_R2_val_2_psorted_dedup_peaks_ucsc_bl.bed
hs_te_93_pks=/home/cmq/bioinfo/project-cmq/embryo_93/results/idr/p_0.05/filtered/hTE93-IP-1_combined_R1_val_1_hTE93-IP-1_combined_R2_val_2_psorted_dedup_peaks_hTE93-IP-2_combined_R1_val_1_hTE93-IP-2_combined_R2_val_2_psorted_dedup_peaks_ucsc_bl.bed
hs_tsc_93_pks=/home/data/wanglab/hTSCs_histone/analysis/htsc/results/idr/broad_p0.05/Homo_sapiens_H3K9me3_BR1_TR1_val_1_Homo_sapiens_H3K9me3_BR1_TR2_val_2_psort_dedup_peaks_Homo_sapiens_H3K9me3_BR2_TR1_val_1_Homo_sapiens_H3K9me3_BR2_TR2_val_2_psort_dedup_peaks_ucsc_bl.bed
hs_esc_93_pks=/home/data/publicdata/GSE62562/analysis/results/idr/broad_p0.05/H3K9me3_XaXe_trimmed_psort_dedup_peaks_H3K9me3_XaXi_trimmed_psort_dedup_peaks_ucsc_bl.bed
hs_fibroblast_93_pks=/home/cmq/bioinfo/project-cmq/gse62562/results/macs2/p_0.05/filtered/SRR1617639_trimmed_psort_dedup_bpks_bl.bed

# - human H3K27me3 peaks
hs_c4_273_pks=/home/cmq/bioinfo/project-cmq/gse124718/results/idr/multi_merged_BAMPE_p_0.05/filtered/SRR8402616_SRR8402616_psorted_dedup_peaks_SRR8402617_SRR8402617_psorted_dedup_peaks_ucsc_bl.bed
hs_c8_273_pks=/home/cmq/bioinfo/project-cmq/gse124718/results/idr/multi_merged_BAMPE_p_0.05/filtered/SRR8402630_SRR8402630_psorted_dedup_SRR8402631_SRR8402631_psorted_dedup_merged_peaks_SRR8402633_SRR8402633_psorted_dedup_SRR8402634_SRR8402634_psorted_dedup_merged_peaks_ucsc_bl.bed
hs_morula_273_pks=/home/cmq/bioinfo/project-cmq/gse124718/results/idr/multi_merged_BAMPE_p_0.05/filtered/SRR8249175_1_val_1_SRR8249175_2_val_2_psort_dedup_peaks_SRR8249176_1_val_1_SRR8249176_2_val_2_psort_dedup_peaks_ucsc_bl.bed
hs_icm_273_pks=/home/cmq/bioinfo/project-cmq/gse124718/results/idr/multi_merged_BAMPE_p_0.05/filtered/SRR8402638_trimmed_psorted_dedup_peaks_SRR8402639_trimmed_psorted_dedup_peaks_ucsc_bl.bed
hs_te_273_pks=/home/cmq/bioinfo/project-cmq/gse124718/results/idr/multi_merged_BAMPE_p_0.05/filtered/SRR9131731_SRR9131731_psorted_dedup_peaks_SRR9131732_SRR9131732_psorted_dedup_peaks_ucsc_bl.bed
hs_tsc_273_pks=/home/data/wanglab/hTSCs_histone/analysis/htsc/results/idr/broad_p0.05/Homo_sapiens_H3K27me3_BR1_TR1_val_1_Homo_sapiens_H3K27me3_BR1_TR2_val_2_psort_dedup_peaks_Homo_sapiens_H3K27me3_BR2_TR1_val_1_Homo_sapiens_H3K27me3_BR2_TR2_val_2_psort_dedup_peaks_ucsc_bl.bed
hs_esc_273_pks=/home/data/publicdata/GSE62562/analysis/results/idr/broad_p0.05/H3K27me3_XaXe_trimmed_psort_dedup_peaks_H3K27me3_XaXi_trimmed_psort_dedup_peaks_ucsc_bl.bed
hs_fibroblast_273_pks=/home/cmq/bioinfo/project-cmq/gse62562/results/macs2/p_0.05/filtered/SRR1617636_trimmed_psort_dedup_bpks_bl.bed

# - mouse H3K9me3 peaks
mm_c2_93_pks=/home/cmq/bioinfo/project-cmq/gse98149_mouse_93/results/idr/0.05/filtered/SRR5479607_1_val_1_SRR5479607_2_val_2_psorted_dedup_peaks_SRR5479608_1_val_1_SRR5479608_2_val_2_psorted_dedup_peaks_ucsc_bl.bed
mm_c4_93_pks=/home/cmq/bioinfo/project-cmq/gse98149_mouse_93/results/idr/0.05/filtered/SRR5479609_1_val_1_SRR5479609_2_val_2_psorted_dedup_peaks_SRR5479610_1_val_1_SRR5479610_2_val_2_psorted_dedup_peaks_ucsc_bl.bed
mm_c8_93_pks=/home/cmq/bioinfo/project-cmq/gse98149_mouse_93/results/idr/0.05/filtered/SRR5479611_1_val_1_SRR5479611_2_val_2_psorted_dedup_peaks_SRR5479612_1_val_1_SRR5479612_2_val_2_psorted_dedup_peaks_ucsc_bl.bed
mm_morula_93_pks=/home/cmq/bioinfo/project-cmq/gse98149_mouse_93/results/idr/0.05/filtered/SRR5479613_1_val_1_SRR5479613_2_val_2_psorted_dedup_peaks_SRR5479614_1_val_1_SRR5479614_2_val_2_psorted_dedup_peaks_ucsc_bl.bed
mm_icm_93_pks=/home/cmq/bioinfo/project-cmq/gse98149_mouse_93/results/idr/0.05/filtered/SRR5479615_1_val_1_SRR5479615_2_val_2_psorted_dedup_peaks_SRR5479616_1_val_1_SRR5479616_2_val_2_psorted_dedup_peaks_ucsc_bl.bed
mm_te_93_pks=/home/cmq/bioinfo/project-cmq/gse98149_mouse_93/results/idr/0.05/filtered/SRR5479617_1_val_1_SRR5479617_2_val_2_psorted_dedup_peaks_SRR5479618_1_val_1_SRR5479618_2_val_2_psorted_dedup_peaks_ucsc_bl.bed

# - mouse H3K27me3 peaks
mm_c2_273_pks=/home/cmq/bioinfo/project-cmq/gse73952_mouse_27/results/idr/p_0.05/filtered/SRR3208757_1_val_1_SRR3208757_2_val_2_psorted_dedup_peaks_SRR3208758_1_val_1_SRR3208758_2_val_2_psorted_dedup_peaks_ucsc_bl.bed
mm_c4_273_pks=/home/cmq/bioinfo/project-cmq/gse73952_mouse_27/results/idr/p_0.05/filtered/SRR3208763_1_val_1_SRR3208763_2_val_2_psorted_dedup_peaks_SRR3208764_1_val_1_SRR3208764_2_val_2_psorted_dedup_peaks_ucsc_bl.bed
mm_c8_273_pks=/home/cmq/bioinfo/project-cmq/gse73952_mouse_27/results/idr/p_0.05/filtered/SRR3208770_1_val_1_SRR3208770_2_val_2_psorted_dedup_peaks_SRR3208771_1_val_1_SRR3208771_2_val_2_psorted_dedup_peaks_ucsc_bl.bed
mm_morula_273_pks=/home/cmq/bioinfo/project-cmq/gse73952_mouse_27/results/idr/p_0.05/filtered/SRR3208776_1_val_1_SRR3208776_2_val_2_psorted_dedup_peaks_SRR3208777_1_val_1_SRR3208777_2_val_2_psorted_dedup_peaks_ucsc_bl.bed
mm_icm_273_pks=/home/cmq/bioinfo/project-cmq/gse73952_mouse_27/results/idr/p_0.05/filtered/SRR3208781_1_val_1_SRR3208781_2_val_2_psorted_dedup_peaks_SRR3208782_1_val_1_SRR3208782_2_val_2_psorted_dedup_peaks_ucsc_bl.bed
mm_te_273_pks=/home/cmq/bioinfo/project-cmq/gse73952_mouse_27/results/idr/p_0.05/filtered/SRR3208786_1_val_1_SRR3208786_2_val_2_psorted_dedup_peaks_SRR3208787_1_val_1_SRR3208787_2_val_2_psorted_dedup_peaks_ucsc_bl.bed

# - human H3K9me3 and H3K27me3 bivalent peaks bed file
c4_93_273_pks=/home/cmq/bioinfo/project-cmq/gse124718/results/idr/multi_merged_BAMPE_p_0.05/filtered/k27_k9/h4C_k93_k27_0.5_ucsc_bl.bed
c8_93_273_pks=/home/cmq/bioinfo/project-cmq/gse124718/results/idr/multi_merged_BAMPE_p_0.05/filtered/k27_k9/h8C_k93_k27_0.5_ucsc_bl.bed
icm_93_273_pks=/home/cmq/bioinfo/project-cmq/gse124718/results/idr/multi_merged_BAMPE_p_0.05/filtered/k27_k9/hICM_k93_k27_0.5_ucsc_bl.bed
morula_93_273_pks=/home/cmq/bioinfo/project-cmq/gse124718/results/idr/multi_merged_BAMPE_p_0.05/filtered/k27_k9/hM_k93_k27_0.5_ucsc_bl.bed
te_93_273_pks=/home/cmq/bioinfo/project-cmq/gse124718/results/idr/multi_merged_BAMPE_p_0.05/filtered/k27_k9/hTE_k93_k27_0.5_ucsc_bl.bed

# - repeat elements and gene annotation file in different format
hs_re_bed=/home/cmq/genome/ucsc/human/hg38/repeat/GRCh38_RepeatMasker_annotation_RepeatRegion_with_strand_ucsc.bed
mm_re_bed=/home/cmq/genome/ucsc/mouse/mm10/repeat/GRCm38_RepeatMasker_annotation_RepeatRegion_with_strand_ucsc.bed
gene_body=/home/cmq/genome/ensembl/release97/homo_sapiens/Homo_sapiens.GRCh38.97.ens2ucsc.all.gene.body.bed
pro500=/home/cmq/genome/ensembl/release97/homo_sapiens/Homo_sapiens.GRCh38.97.ens2ucsc.all.gene.promoter.500bp.bed
repeat=/home/cmq/genome/ucsc/human/hg38/repeat/GRCh38_RepeatMasker_annotation_RepeatRegion_ucsc.bed
repeat2=/home/cmq/genome/ucsc/human/hg38/repeat/GRCh38_RepeatMasker_repeat_anno_quan.bed

# - H3K9me3 bigwig files
c8_input=/home/cmq/bioinfo/project-cmq/embryo_93/results/deeptools/bigwig/merged_bam/h8C93-Input-2_combined_R1_val_1_h8C93-Input-2_combined_R2_val_2_psorted_dedup.bw
c8=/home/cmq/bioinfo/project-cmq/embryo_93/results/deeptools/bigwig/merged_bam/h8C93-Ip-3_combined_R1_val_1_h8C93-Ip-3_combined_R2_val_2_psorted_dedup_h8C93-Ip-4_combined_R1_val_1_h8C93-Ip-4_combined_R2_val_2_psorted_dedup_merged.bw
icm_input=/home/cmq/bioinfo/project-cmq/embryo_93/results/deeptools/bigwig/merged_bam/hICM93-Input_combined_R1_val_1_hICM93-Input_combined_R2_val_2_psorted_dedup.bw
icm=/home/cmq/bioinfo/project-cmq/embryo_93/results/deeptools/bigwig/merged_bam/hICM93-IP-1_combined_R1_val_1_hICM93-IP-1_combined_R2_val_2_psorted_dedup_hICM93-IP-2_combined_R1_val_1_hICM93-IP-2_combined_R2_val_2_psorted_dedup_merged.bw

# - H3K27ac bigwig files
c8_27ac_rep1=/home/yhw/bioinfo/project-temp/wjc/gse124718/results/deeptools/coverage/SRR9131735_1_val_1_SRR9131735_2_val_2_psort_dedup.bw
c8_27ac_rep2=/home/yhw/bioinfo/project-temp/wjc/gse124718/results/deeptools/coverage/SRR9131736_1_val_1_SRR9131736_2_val_2_psort_dedup.bw
icm_27ac_rep1=/home/yhw/bioinfo/project-temp/wjc/gse124718/results/deeptools/coverage/SRR9131737_1_val_1_SRR9131737_2_val_2_psort_dedup.bw
icm_27ac_rep2=/home/yhw/bioinfo/project-temp/wjc/gse124718/results/deeptools/coverage/SRR9131738_1_val_1_SRR9131738_2_val_2_psort_dedup.bw

# - ATAC-seq bigwig files
c8_atac_rep1=/home/cmq/bioinfo/project-ymz/atac/gse101571/results/deeptools/bigwig/dedup_merge_bam/SRR5837333_1_val_1_SRR5837333_2_val_2_psorted_dedup_SRR5837334_1_val_1_SRR5837334_2_val_2_psorted_dedup_merged.bw
c8_atac_rep2=/home/cmq/bioinfo/project-ymz/atac/gse101571/results/deeptools/bigwig/dedup_merge_bam/SRR5837335_1_val_1_SRR5837335_2_val_2_psorted_dedup_SRR5837336_1_val_1_SRR5837336_2_val_2_psorted_dedup_merged.bw
icm_atac_rep1=/home/cmq/bioinfo/project-ymz/atac/gse101571/results/deeptools/bigwig/dedup_merge_bam/SRR5837337_1_val_1_SRR5837337_2_val_2_psorted_dedup_SRR5837338_1_val_1_SRR5837338_2_val_2_psorted_dedup_merged.bw
icm_atac_rep2=/home/cmq/bioinfo/project-ymz/atac/gse101571/results/deeptools/bigwig/dedup_merge_bam/SRR5837339_1_val_1_SRR5837339_2_val_2_psorted_dedup_SRR5837340_1_val_1_SRR5837340_2_val_2_psorted_dedup_SRR5837341_1_val_1_SRR5837341_2_val_2_psorted_dedup_merged.bw

# - H3K4me3 bigwig files
c8_43=/home/cmq/bioinfo/project-cmq/gse124718_human_43/results/deeptools/bigwig/merged_bam/SRR8402624_1_val_SRR8402625_1_val_SRR8402627_1_val_SRR8402628_1_valmerged.bw
icm_43=/home/cmq/bioinfo/project-cmq/gse124718_human_43/results/deeptools/bigwig/merged_bam/SRR8402635_trimmed_psorted_dedup_SRR8402636_trimmed_psorted_dedup_merged.bw

# - H3K27me3 bigwig files
c8_27me3=/home/cmq/bioinfo/project-cmq/gse124718/results/deeptools/bigwig/merge_replicate_multi/8cell_multi_replicate_merged.bw
icm_27me3=/home/cmq/bioinfo/project-cmq/gse124718/results/deeptools/bigwig/merge_replicate_multi/icm_multi_replicate_merged.bw

# - promoter regions of genes contained H3K9me3 and H3K27me3 bivalent L1/SVA
bi_l1_pro=/home/cmq/bioinfo/project-cmq/embryo_93/results/idr/p_0.05/filtered/covered_repeat_age/seperate/h8C_k93_k27_0.5_ucsc_bl_covered_L1HS_L1PA2_L1PA3_quan_50pect_marked_gene_low_pro500bp.bed
bi_sva_pro=/home/cmq/bioinfo/project-cmq/embryo_93/results/idr/p_0.05/filtered/covered_repeat_age/seperate/h8C_k93_k27_0.5_ucsc_bl_covered_SVA_D_SVA_F_quan_50pect_marked_gene_low_pro500bp.bed



### >>> Calculate human/mouse H3K9me3/H3K27me3 signal on full length L1 rgeions through homemade shell scripts (CMQ)
wd=/home/yhw/bioinfo/project-mine/Embryo.93/chipseq
nohup bash ${wd}/hs_k9me3/scripts/BeautyMapByHand-1.sh -c ${wd}/hs_k9me3/scripts/BeautyMapByHand-1c_fl_LINE1.txt &
nohup bash ${wd}/hs_k27me3/scripts/BeautyMapByHand-1.sh -c ${wd}/hs_k27me3/scripts/BeautyMapByHand-1c_fl_LINE1.txt &
nohup bash ${wd}/mm_k9me3/scripts/BeautyMapByHand-1.sh -c ${wd}/mm_k9me3/scripts/BeautyMapByHand-1c_fl_LINE1.txt &
nohup bash ${wd}/mm_k27me3/scripts/BeautyMapByHand-1.sh -c ${wd}/mm_k27me3/scripts/BeautyMapByHand-1c_fl_LINE1.txt &



### >>> Human H3K9me3 and H3K27me3 bivalent repeat annotation (CMQ)
# - H3K9me3 and H3K27me3 overlapped peaks
od=/home/cmq/bioinfo/project-cmq/embryo_93/results/idr/p_0.05/filtered/covered_repeat_age/seperate
# - all stages
bedtools intersect -a ${hs_c4_273_pks} -b ${hs_c4_93_pks} -f 0.5 > ${od}/h4C_k93_k27_0.5_ucsc_bl.bed 
bedtools intersect -a ${hs_c8_273_pks} -b ${hs_c8_93_pks} -f 0.5 > ${od}/h8C_k93_k27_0.5_ucsc_bl.bed
bedtools intersect -a ${hs_morula_273_pks} -b ${hs_morula_93_pks} -f 0.5 > ${od}/hM_k93_k27_0.5_ucsc_bl.bed
bedtools intersect -a ${hs_icm_273_pks} -b ${hs_icm_93_pks} -f 0.5 > ${od}/hICM_k93_k27_0.5_ucsc_bl.bed
bedtools intersect -a ${hs_te_273_pks} -b ${hs_te_93_pks} -f 0.5 > ${od}/hTE_k93_k27_0.5_ucsc_bl.bed
bedtools intersect -a ${hs_fibroblast_273_pks} -b ${hs_fibroblast_93_pks} -f 0.5 > /home/cmq/bioinfo/project-cmq/gse62562/results/macs2/p_0.05/filtered/covered_repeat_age/fibroblast_93_27_shared_peaks_bl.bed
bedtools intersect -a ${hs_esc_273_pks} -b ${hs_esc_93_pks} -f 0.5 > /home/data/publicdata/GSE62562/analysis/results/idr/broad_p0.05/H3K27me3_H3K9me3_hESC_shared_peaks_ucsc_bl.bed
bedtools intersect -a ${hs_tsc_273_pks} -b ${hs_tsc_93_pks} -f 0.5 > /home/data/wanglab/hTSCs_histone/analysis/htsc/results/idr/broad_p0.05/H3K27me3_H3K9me3_hTSC_shared_peaks_ucsc_bl.bed
# - cleavage
cat ${od}/h4C_k93_k27_0.5_ucsc_bl.bed ${od}/h8C_k93_k27_0.5_ucsc_bl.bed | sort -k1,1 -k2,2n \
    | bedtools merge -i stdin > ${od}/c4_c8_93_27_shared_peaks_merged_ucsc_bl.bed
# - blastocyst
cat ${od}/hICM_k93_k27_0.5_ucsc_bl.bed ${od}/hTE_k93_k27_0.5_ucsc_bl.bed | sort -k1,1 -k2,2n \
    | bedtools merge -i stdin > ${od}/icm_te_93_27_shared_peaks_merged_ucsc_bl.bed
# - K9-K27 bivalent repeat annotation
cat ${hs_re_bed} | awk '{if($5 == "LINE" || $5 == "LTR" || $5 == "Retroposon") print $0}' \
    | cut -f 7 | sort | uniq -c | awk '{print $1"\t"$2}' > ${od}/filtered_locus_num_on_subfamily.txt
cat ${hs_re_bed} | awk '{if($5 == "LINE" || $5 == "LTR" || $5 == "Retroposon") print $7"\t"$5"\t"$6}' \
    | sort | uniq > ${od}/filtered_locus_anno.txt
ls ${od}/*merged_ucsc_bl.bed | while read bed
do
   filename=${bed##*/}; pre=${filename%%.*}
   bedtools intersect -a ${hs_re_bed} -b ${bed} -f 0.5 > ${od}/${pre}_covered_repeat_50pect.bed
   cat ${od}/${pre}_covered_repeat_50pect.bed | awk '{if($5 == "LINE" || $5 == "LTR" || $5 == "Retroposon") print $0}' \
       | cut -f 7 | sort | uniq -c | sort -k 1,1nr | awk '{print $1"\t"$2}' > ${od}/${pre}_covered_repeat_50pect_tar_subfamily_num.txt
   cat ${od}/${pre}_covered_repeat_50pect.bed | awk '{if($5 == "LINE" || $5 == "LTR" || $5 == "Retroposon") print $0}' \
       | csvtk summary -f 8:mean -g 7 -t -H > ${od}/${pre}_covered_repeat_50pect_tar_age.txt
   cat ${od}/${pre}_covered_repeat_50pect.bed | awk '{if($5 == "LINE" || $5 == "LTR" || $5 == "Retroposon") print $7"\t"$3-$2}' \
       | csvtk summary -H -f 2:mean -g 1 -t > ${od}/${pre}_covered_repeat_50pect_tar_mean_length_on_subfamily.txt
   csvtk join -H -t -f "1;2;2;1;1" ${od}/${pre}_covered_repeat_50pect_tar_age.txt ${od}/${pre}_covered_repeat_50pect_tar_subfamily_num.txt \
       ${od}/filtered_locus_num_on_subfamily.txt ${od}/${pre}_covered_repeat_50pect_tar_mean_length_on_subfamily.txt \
       ${od}/filtered_locus_anno.txt > ${od}/${pre}_covered_repeat_50pect_tar.txt
done


### >>> Mouse H3K9me3 and H3K27me3 bivalent repeat annotation (CMQ)
# - H3K9me3 and H3K27me3 overlapped peaks
od=/home/cmq/bioinfo/project-cmq/gse73952_mouse_27/results/idr/p_0.05/filtered/k27_k9/covered_repeat_age
# - all stage 
bedtools intersect -a ${mm_c2_273_pks} -b ${mm_c2_93_pks} -f 0.5 > ${od}/m2c_k93_k27_ucsc_bl.bed
bedtools intersect -a ${mm_c4_273_pks} -b ${mm_c4_93_pks} -f 0.5 > ${od}/m4c_k93_k27_ucsc_bl.bed
bedtools intersect -a ${mm_c8_273_pks} -b ${mm_c8_93_pks} -f 0.5 > ${od}/m8c_k93_k27_ucsc_bl.bed
bedtools intersect -a ${mm_morula_273_pks} -b ${mm_morula_93_pks} -f 0.5 > ${od}/morula_k93_k27_ucsc_bl.bed
bedtools intersect -a ${mm_icm_273_pks} -b ${mm_icm_93_pks} -f 0.5 > ${od}/icm_k93_k27_ucsc_bl.bed
bedtools intersect -a ${mm_te_273_pks} -b ${mm_te_93_pks} -f 0.5 > ${od}/te_k93_k27_ucsc_bl.bed
# - blastocyst
cat ${od}/icm_k93_k27_ucsc_bl.bed ${od}/te_k93_k27_ucsc_bl.bed > ${od}/icm_te_k93_k27_merged_ucsc_bl.bed
# - K9-K27 bivalent repeat annotation
# 2Cell,4Cell and 8Cell (annotation via homemade shell script separately)
nohup bash ${od}/../AnnoPksByRepeat.sh -c ${od}/../AnnoPksByRepeat.para.txt &
# ICM and TE
ls ${od}/*merged_ucsc_bl.bed | while read bed 
do
   filename=${bed##*/}; prefix=${filename%.*}
   bedtools intersect -a ${mm_re_bed} -b ${bed} -f 0.5 > ${od}/${prefix}_covered_repeat_50pect.bed
   cat ${od}/${prefix}_covered_repeat_50pect.bed | awk '{if($5 == "LINE" || $5 == "LTR") print $0}' | cut -f 7 | sort | uniq -c | sort -k 1,1nr \
       | awk '{print $1"\t"$2}' > ${od}/${prefix}_covered_repeat_50pect_tar_subfamily_num.txt
   cat ${od}/${prefix}_covered_repeat_50pect.bed | awk '{if($5 == "LINE" || $5 == "LTR") print $0}' \
       | csvtk summary -f 8:mean -g 7 -t -H > ${od}/${prefix}_covered_repeat_50pect_tar_age.txt
   cat ${od}/${prefix}_covered_repeat_50pect.bed | awk '{if($5 == "LINE" || $5 == "LTR") print $7"\t"$3-$2}' \
       | csvtk summary -H -f 2:mean -g 1 -t > ${od}/${prefix}_covered_repeat_50pect_tar_mean_length_on_subfamily.txt
   csvtk join -H -t -f "1;2;2;1;1" ${od}/${prefix}_covered_repeat_50pect_tar_age.txt \
       ${od}/${prefix}_covered_repeat_50pect_tar_subfamily_num.txt ${od}/filtered_locus_num_on_subfamily.txt \
       ${od}/${prefix}_covered_repeat_50pect_tar_mean_length_on_subfamily.txt ${od}/filtered_locus_anno.txt > ${od}/${prefix}_covered_repeat_50pect_tar.txt
done



### >>> Number of human H3K9me3 and H3K27me3 bivalent repeats, genes and promoters (CMQ)
od=/home/cmq/bioinfo/project-cmq/gse124718/results/idr/multi_merged_BAMPE_p_0.05/filtered/k27_k9/covered_gene_repeat/
# - marked repeats
for stage in c4_93_273_pks c8_93_273_pks icm_93_273_pks morula_93_273_pks te_93_273_pks
do 
    num=`bedtools intersect -a ${repeat} -b ${!stage} -f 0.5 -wa \
             | bedtools intersect -v -a stdin -b ${gene_body} \
             | bedtools intersect -v -a stdin -b ${pro500} | wc -l`
    echo -e "${stage}\t${num}\trepeat" >> ${od}/H3K9me3_H3K27me3_shared_peaks_marked_genebody_promoter500bp_repeat.txt
done
# - marked promoters
for stage in c4_93_273_pks c8_93_273_pks icm_93_273_pks morula_93_273_pks te_93_273_pks
do
    num=`cat ${pro500} | awk '{if($6=="protein_coding") print $0}' | bedtools intersect -a stdin -b ${!stage} -f 0.5 -wa | wc -l`
    echo -e "${stage}\t${num}\tpromoter" >> ${od}/H3K9me3_H3K27me3_shared_peaks_marked_genebody_promoter500bp_repeat.txt
done
# - marked gene body regions
for stage in c4_93_273_pks c8_93_273_pks icm_93_273_pks morula_93_273_pks te_93_273_pks
do
    num=`cat ${gene_body} | awk '{if($6=="protein_coding") print $0}' | bedtools intersect -a stdin -b ${!stage} -f 0.5 -wa | wc -l`
    echo -e "${stage}\t${num}\tgenebody" >> ${od}/H3K9me3_H3K27me3_shared_peaks_marked_genebody_promoter500bp_repeat.txt
done


### >>> Chromatin state around TSS of genes contained H3K9me3-HK27me3 bivalent L1/SVA (CMQ)

# - 8Cell H3K9me3-H3K27me3 bivalent L1HS, L1PA2, L1PA3, SVA_D and SVA_F marked gene
wd=/home/cmq/bioinfo/project-cmq/embryo_93/results/idr/p_0.05/filtered/covered_repeat_age/seperate
# L1HS,L1PA2,L1PA3
bedtools intersect -a ${repeat2} -b ${c8_93_273_pks} -f 0.5 -wa | grep "L1HS\|L1PA2\|L1PA3" \
    | cut -f 1-4 | grep -v "fix" | grep -v "alt" > ${wd}/h8C_k93_k27_0.5_ucsc_bl_covered_L1HS_L1PA2_L1PA3_quan_50pect.bed
cat ${wd}/h8C_k93_k27_0.5_ucsc_bl_covered_L1HS_L1PA2_L1PA3_quan_50pect.bed \
    | awk '{if(($3-$2)>1000) print $0}' \
    | bedtools intersect -a stdin -b ${gene_body} -f 0.9 -wa -wb | cut -f 8 | sort | uniq \
    > ${wd}/h8C_k93_k27_0.5_ucsc_bl_covered_L1HS_L1PA2_L1PA3_quan_50pect_marked_gene.txt
# SVA_D,SVA_F
bedtools intersect -a ${repeat2} -b ${c8_93_273_pks} -f 0.5 -wa | grep "SVA_D\|SVA_F" \
         | cut -f 1-4 | grep -v "fix" | grep -v "alt" > ${wd}/h8C_k93_k27_0.5_ucsc_bl_covered_SVA_D_SVA_F_quan_50pect.bed
cat ${wd}/h8C_k93_k27_0.5_ucsc_bl_covered_SVA_D_SVA_F_quan_50pect.bed \
    | awk '{if(($3-$2)>1000) print $0}' \
    | bedtools intersect -a stdin -b ${gene_body} -f 0.9 -wa -wb | cut -f 8 | sort | uniq \
    > ${wd}/h8C_k93_k27_0.5_ucsc_bl_covered_SVA_D_SVA_F_quan_50pect_marked_gene.txt

# - Promoters of 8Cell H3K9me3-H3K27me3 bivalent L1HS, L1PA2, L1PA3, SVA_D and SVA_F marked gene with low expression level
grep -w -f ${wd}/h8C_k93_k27_0.5_ucsc_bl_covered_L1HS_L1PA2_L1PA3_quan_50pect_marked_gene_low.txt ${pro500} | cut -f 1-4 \
    > ${wd}/h8C_k93_k27_0.5_ucsc_bl_covered_L1HS_L1PA2_L1PA3_quan_50pect_marked_gene_low_pro500bp.bed
grep -w -f ${wd}/h8C_k93_k27_0.5_ucsc_bl_covered_SVA_D_SVA_F_quan_50pect_marked_gene_low.txt ${pro500} | cut -f 1-4 \
    > ${wd}/h8C_k93_k27_0.5_ucsc_bl_covered_SVA_D_SVA_F_quan_50pect_marked_gene_low_pro500bp.bed
cat ${wd}/h8C_k93_k27_0.5_ucsc_bl_covered_L1HS_L1PA2_L1PA3_quan_50pect_marked_gene_low_pro500bp.bed \
    ${wd}/h8C_k93_k27_0.5_ucsc_bl_covered_SVA_D_SVA_F_quan_50pect_marked_gene_low_pro500bp.bed \
    > ${wd}/h8C_k93_k27_0.5_ucsc_bl_covered_L1HS_L1PA2_L1PA3_SVA_D_SVA_F_quan_50pect_marked_gene_low_pro500bp.bed

# - Visualize H3K9me3, H3K27me3, H3K27ac, H3K4me3 and ATAC-seq signal around TSS of genes comtained K9-K27 bi-L1/SVA
wd=/home/cmq/bioinfo/project-cmq/embryo_93
mkdir -p ${wd}/results/deeptools/matrix/c8_bi_l1_sva_ge_low
mkdir -p ${wd}/results/deeptools/plotprofile/c8_bi_l1_sva_ge_low
computeMatrix reference-point -S ${c8_input} ${c8} ${c8_27me3} ${c8_27ac_rep1} ${c8_27ac_rep2} ${c8_43} ${c8_atac_rep1} ${c8_atac_rep2} \
                              -R ${bi_l1_pro} ${bi_sva_pro} --averageTypeBins mean --referencePoint center \
                              --beforeRegionStartLength 3000 --afterRegionStartLength 3000 --numberOfProcessors 12 --missingDataAsZero \
                              -o ${wd}/results/deeptools/matrix/c8_bi_l1_sva_ge_low/c8_ref_gzmatrix.gz

plotProfile -m ${wd}/results/deeptools/matrix/c8_bi_l1_sva_ge_low/c8_ref_gzmatrix.gz \
            -o ${wd}/results/deeptools/plotprofile/c8_bi_l1_sva_ge_low/human_8Cell_H3K27ac_ATAC_H3K4me3_signal_on_8Cell_bivalent_H3K27me3_and_H3K9me3_L1_SVA_marked_gene_low_promoter500_line.pdf \
            --averageType mean --plotType lines --legendLocation upper-left --plotHeight 10 --plotWidth 12 \
            --samplesLabel "H3K9me3_Input" "H3K9me3" "H3K27me3" "H3K27ac_1" "H3K27ac_2" "H3K4me3" "ATAC_1" "ATAC_2" --perGroup \
            --regionsLabel "Bi_L1" "Bi_SVA"

computeMatrix reference-point -S ${icm_input} ${icm} ${icm_27me3} ${icm_27ac_rep1} ${icm_27ac_rep2} ${icm_43} ${icm_atac_rep1} ${icm_atac_rep2} \
                              -R ${bi_l1_pro} ${bi_sva_pro} --averageTypeBins mean --referencePoint center \
                              --beforeRegionStartLength 3000 --afterRegionStartLength 3000 --numberOfProcessors 12 --missingDataAsZero \
                              -o ${wd}/results/deeptools/matrix/c8_bi_l1_sva_ge_low/icm_ref_gzmatrix.gz

plotProfile -m ${wd}/results/deeptools/matrix/c8_bi_l1_sva_ge_low/icm_ref_gzmatrix.gz \
            -o ${wd}/results/deeptools/plotprofile/c8_bi_l1_sva_ge_low/human_ICM_H3K27ac_ATAC_H3K4me3_signal_on_8Cell_bivalent_H3K27me3_and_H3K9me3_L1_SVA_marked_gene_low_promoter500_line.pdf \
            --averageType mean --plotType lines --legendLocation upper-left --plotHeight 10 --plotWidth 12 \
            --samplesLabel "H3K9me3_Input" "H3K9me3" "H3K27me3" "H3K27ac_1" "H3K27ac_2" "H3K4me3" "ATAC_1" "ATAC_2" --perGroup \
            --regionsLabel "Bi_L1" "Bi_SVA"

